rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,482 @@
+# RFC#1440: Configuration-based Generic Planner Node
+
+## Objective
+
+Make the node graph easier to set up and reason about, and amenable to static
+analysis.
+
+## Background
+
+An Oak application creates an initial node and calls the entry point on that
+node. A typical pattern is to use this initial node as a "Planner Node",
+responsible for creating all other nodes and channels. The creation of the nodes
+and channels is done through code. For simple cases it is easy to figure out
+what the graph can look like by inspecting the code, but it becomes increasingly
+difficult as the complexity of the logic of the planner node increases to cope
+with more complex node graphs. Another drawback of the code-based dynamic node
+creation is that we have to run the code to use the introspection mechanism to
+visualise the node graph.
+
+Generating the node graph automatically based on a configuration/manifest file
+should simplify this. It should be easier to figure out what the node graph will
+look like through inspecting a manifest file (assuming it has an easily
+understandable format). Additional benefits of configuration-based node creation
+include:
+
+- It would be easier to build tools to visualise the node graph without running
+  the application.
+- It would be easier to build tools to perform static analysis of the node graph
+  to detect potential problems.
+- It would usually be easier to reason about the security of a static node
+  graph.
+
+## Overview
+
+Moving all node and channel creation to a configuration-based mechanism is a
+significant and potentially disruptive change. This RFC proposes a more gradual
+approach by implementing a re-usable Wasm-based planner node that uses a
+manifest file to do the node and channel creation. The functionality can be
+expanded over time to support more complex scenarios.
+
+The planner node will be responsible for creating nodes, creating channels,
+sending channel handles to the created nodes to ensure that the full node graph
+is created on start-up. It will also act as a message router, which would route
+HTTP- and GRPC invocations to other nodes based on the associated labels.
+
+Some applications require dynamic creation of nodes and channels based on the
+labels associated with incoming invocations. A sub-graph template mechanism will
+be implemented to support dynamic node creation. The typical scenario where this
+is required is the case where a per-user node must be created to support
+per-user labels. Although this means that the entire node graph cannot be
+created up-front and that the node graph will not be static, it should still be
+easier to reason about each of the sub-graphs and the overall node graph is
+still strongly constrained.
+
+## Detailed Design
+
+### Planner Node
+
+The planner node will be the initial node created for an application. The
+manifest file will be passed to the planner node as part of the application's
+`ConfigMap`.
```

Comment:
How would a client specify such Wasm hash label? Also it was not clear to me that this planner node is supposed to create things that are "safer" than arbitrary Wasm logic, could you clarify that?

Perhaps this is part of the line above:

> - It would usually be easier to reason about the security of a static node
  graph.

But it would be good to have an actual example of something that is insecure in Wasm, and how the planner node prevents / solves that.